core/src/storage2/collections/binary_heap/elements.rs

```diff
@@ -0,0 +1,291 @@
+    }
+
+    /// Returns an exclusive reference to the first element if any.
+    pub fn first_mut(&mut self) -> Option<&mut T> {
+        if self.is_empty() {
+            return None
+        }
+        self.get_mut(0)?.as_mut()
+    }
+
+    /// Removes all elements from this vector.
+    ///
+    /// # Note
+    ///
+    /// Use this method to clear the vector instead of e.g. iterative `pop()`.
+    /// This method performs significantly better and does not actually read
+    /// any of the elements (whereas `pop()` does).
+    pub fn clear(&mut self) {
+        if self.is_empty() {
+            return
+        }
+        self.children.clear();
+        self.len = Lazy::new(0);
+    }
+
+    /// Appends an element to the back of the vector.
+    pub fn push(&mut self, value: T) {
+        assert!(
+            self.len() < core::u32::MAX,
+            "cannot push more elements into the storage vector"
+        );
+        let last_index = self.len();
+        *self.len += 1;
+        self.push_to(last_index, Some(value));
+    }
+
+    fn push_to(&mut self, index: u32, value: Option<T>) {
+        let children_index = children::get_children_storage_index(index);
+        match self.children.get_mut(children_index) {
+            Some(children) => {
+                let child_pos = children::get_child_pos(index);
+                *children.child_mut(child_pos) = value;
+            }
+            None => {
+                self.children.push(Children::new(value, None));
+                debug_assert!(
+                    self.children.get(children_index).is_some(),
+                    "the new children were not placed at children_index!"
+                );
+            }
+        };
+    }
+
+    /// Pops the last element from the vector and returns it.
+    //
+    /// Returns `None` if the vector is empty.
+    pub fn pop(&mut self) -> Option<T> {
+        if self.is_empty() {
+            return None
+        }
+        let last_index = self.len() - 1;
+        *self.len = last_index;
+
+        let children_index = children::get_children_storage_index(last_index);
+        let old = self.children.get_mut(children_index);
+        match old {
+            Some(children) => {
+                let child_pos = children::get_child_pos(last_index);
+                let popped_val = children.child_mut(child_pos).take();
+                if children.is_empty() {
+                    // if both children are non-existent the entire children object can be removed
+                    self.children.pop();
+                }
+                popped_val
+            }
+            None => {
+                unreachable!("vec must contain children at children_index of last_index")
+            }
+        }
+    }
+
+    /// Returns the index if it is within bounds or `None` otherwise.
+    fn within_bounds(&self, index: u32) -> Option<u32> {
+        if index < self.len() {
+            return Some(index)
+        }
+        None
+    }
```

Comment:
API docs are not sufficient to state how it shall be used. Above one usage examples is:
```rust
        let storage_index = children::get_children_storage_index(index);
        let i = self.within_bounds(storage_index)?;
```
So I think it should only be used with indices coming from `children::get_children_storage_index` making me wonder if we shouldn't use some proper new-type wrapper instead of a plain old `u32` to make this type safe and implicitly document the usage.